### PR TITLE
Feat/600 files folders

### PR DESCRIPTION
**Pull request by yashksaini-coder**
## Description

This PR introduces **comprehensive and secure file & folder upload/download capabilities** to the Nixopus file manager. The implementation focuses on robustness, security, and backward compatibility, laying the groundwork for upcoming UI integration.

### Backend Enhancements

- **Single File Download**
  - `GET /v1/file-manager/download`
  - Proper content-type detection and HTTP headers

- **Folder Download as ZIP**
  - `GET /v1/file-manager/download-folder`
  - Preserves full directory structure inside ZIP archive
 
---

## Key Features

- **Secure Downloads**:- Authenticated and authorized access to files and folders.

- **ZIP Folder Downloads**:- Download entire directories as compressed ZIP archives.

- **Folder-Aware Uploads**:- Upload nested folder structures while preserving hierarchy.

- **Strong Security Guarantees**:- Protection against directory traversal and invalid inputs.

- **Backward Compatibility**:- Existing single-file upload workflows remain unchanged.

---

## Scope of Change

- [ ] View (UI/UX)
- [x] API
- [ ] CLI
- [ ] Infra / Deployment
- [x] Docs
- [ ] Other: Backend-only changes

---

## Security Considerations

- All endpoints protected via existing **RBAC middleware**
- Centralized **path traversal validation**
- Sanitized inputs for file/folder names
- Safe SFTP connection lifecycle management

---

## Developer Checklist

- [x] PR title is clear and relevant  
- [x] Self-review completed  
- [x] Manual testing performed  
- [x] No secrets or sensitive data exposed  
- [x] No merge conflicts  
- [x] Docs updated (where applicable)  
- [x] Debug prints removed  
- [x] Unit & integration tests passing  
- [x] Follows Nixopus documentation standards  

---

## Technical Implementation Details

```http
GET /v1/file-manager/download?path=<file_path>
GET /v1/file-manager/download-folder?path=<folder_path>
````

### ⬆ Enhanced Upload API

**Endpoint:**
`POST /v1/file-manager/upload`

**New features present:**

* Accepts multiple files via `files[]`
* Supports relative paths via `relativePaths[]`
* Automatically creates directories
* Enforces strict path validation

---

## Add Proper Tests

* Directory traversal attack prevention
* File and folder download workflows
* Upload validation and hierarchy preservation
* Error handling for invalid paths and edge cases

```

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Streamed downloads for single files and folders (folders delivered as ZIP), with safe Content-Disposition handling.
  * Multi-file uploads preserving relative paths, with base-path awareness and automatic directory creation.
  * New download endpoints for file and folder retrieval.

* **Bug Fixes / Security**
  * Path traversal protections and stricter filename sanitization for uploads and downloads.

* **Tests**
  * Added tests for downloads, uploads, and filename sanitization.

* **Documentation**
  * Expanded File Operations and API reference.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

File: api/internal/features/file-manager/controller/download_file.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+func (c *FileManagerController) DownloadFile(f fuego.ContextNoBody) (*shared_types.Response, error) {
+	path := f.QueryParam("path")
+	if path == "" {
+		return nil, fuego.HTTPError{
+			Err:    fuego.NewError(http.StatusBadRequest, "path query parameter is required"),
+			Status: http.StatusBadRequest,
+		}
+	}
+
+	// Path validation to prevent directory traversal
```
Comment (raghavyuva) [inline api/internal/features/file-manager/controller/download_file.go:24]: Any specific reason why absolute path traversal is prevented from here?

File: api/internal/features/file-manager/controller/download_folder.go
Comment (raghavyuva) [inline api/internal/features/file-manager/controller/download_folder.go:1]: if absolute path detection, setting content headers for download, relative paths are performed frequently on two or more files then it is better to generalize them and make use as utils

File: api/internal/features/file-manager/controller/filename_utils.go
Comment (raghavyuva) [inline api/internal/features/file-manager/controller/filename_utils.go:1]: can we split this into smaller chunks of funcs? so it would be easy to understand, i prefer it to be lesser condition blocks and more functions so clear intention and SRP will be followed

File: api/internal/features/file-manager/controller/upload_file.go [hunk 2]
```diff
@@ -9,32 +9,103 @@ import (
 )
 
 func (c *FileManagerController) UploadFile(f fuego.ContextNoBody) (*shared_types.Response, error) {
-	file, header, err := f.Request().FormFile("file")
+	// Parse multipart form
+	err := f.Request().ParseMultipartForm(32 << 20) // 32 MB max memory
```
Comment (raghavyuva) [inline api/internal/features/file-manager/controller/upload_file.go:13]: this is interesting part, we need to make this available as a configuration block in user settings, so max file upload can be set, 

you can keep 32mb max as default, but it can be fragile when the code is shipped for production,

File: api/internal/features/file-manager/controller/upload_file.go [hunk 3]
```diff
@@ -9,32 +9,103 @@ import (
...
-	if err != nil {
-		c.logger.Log(logger.Error, err.Error(), "")
-		return nil, fuego.HTTPError{
-			Err:    err,
-			Status: http.StatusInternalServerError,
+	uploadedCount := 0
+
+	// Handle multiple files
+	files := f.Request().MultipartForm.File["files"]
+	if len(files) == 0 {
```
Comment (raghavyuva) [inline api/internal/features/file-manager/controller/upload_file.go:31]: nested if conditions are hard to analyze here, let's refactor this to prevent nested conditons and small focused functions

File: api/internal/features/file-manager/service/manager.go [hunk 4]
```diff
@@ -22,6 +25,13 @@ type FileInfo struct {
 	IsHidden bool      `json:"is_hidden"`
 }
 
+type FileDownload struct {
```
Comment (raghavyuva) [inline api/internal/features/file-manager/service/manager.go:28]: let's move this out to types.go file if it exists already

File: api/internal/features/file-manager/service/manager.go [hunk 5]
```diff
@@ -203,20 +213,39 @@ func (f *FileManagerService) MoveDirectory(fromPath string, toPath string) error
...
 
-// UploadFile handles file upload to the specified path
-func (f *FileManagerService) UploadFile(file io.Reader, path string, filename string) error {
-	f.logger.Log(logger.Info, "uploading file", filename)
+// UploadFile handles file upload to the specified path with support for relative paths in filename
+func (f *FileManagerService) UploadFile(file io.Reader, basePath string, relativePath string) error {
+	f.logger.Log(logger.Info, "uploading file", relativePath)
+
+	// Validate relative path to prevent path traversal attacks
+	if strings.Contains(relativePath, "..") {
```
Comment (raghavyuva) [inline api/internal/features/file-manager/service/manager.go:221]: this path traversal detection is performed multiple times across the PR, is there any reason for the same?

File: api/internal/features/file-manager/service/manager.go [hunk 6]
```diff
@@ -203,20 +213,39 @@ func (f *FileManagerService) MoveDirectory(fromPath string, toPath string) error
...
+		// Ensure the full path is still within the base path
+		rel, err := filepath.Rel(basePath, fullPath)
+		if err != nil || strings.HasPrefix(rel, "..") {
+			return fmt.Errorf("invalid path: attempts to access outside base directory")
 		}
-		targetPath := filepath.Join(path, filename)
-		out, err := client.Create(targetPath)
+
+		// Create all necessary directories
+		dirPath := filepath.Dir(fullPath)
```
Comment (raghavyuva) [inline api/internal/features/file-manager/service/manager.go:237]: we might have to ensure permissions during the controller level so we can throw an early error

File: api/internal/features/file-manager/service/manager.go [hunk 7]
```diff
@@ -231,7 +260,170 @@ func (f *FileManagerService) UploadFile(file io.Reader, path string, filename st
...
+			return fmt.Errorf("path %s is a directory, not a file", path)
+		}
+
+		file, err := client.Open(path)
+		if err != nil {
+			return fmt.Errorf("failed to open file %s: %w", path, err)
+		}
+
+		// Determine content type based on file extension
+		contentType := "application/octet-stream"
```
Comment (raghavyuva) [inline api/internal/features/file-manager/service/manager.go:283]: this block of code is fragile and prone to error, can we have a look at an alternative way that ensures we always send the proper extension regardless?

**Review by raghavyuva — CHANGES_REQUESTED** [review]
Hey @yashksaini-coder, I’ve added a few comments. Could you take a look and address them when you get a chance?

Thanks!

File: api/internal/features/file-manager/controller/download_file.go (hunk 1 above)
Comment (yashksaini-coder) [reply api/internal/features/file-manager/controller/download_file.go:24]: THe `DownloadFile` handles HTTP requests to download a file from the server.
It expects a "path" query parameter specifying the relative path to the file.
The function performs several security checks:
  - Ensures the "path" parameter is provided.
  - Rejects absolute paths to prevent access outside the intended directory structure.
  - Cleans and validates the path to prevent directory traversal attacks (e.g., using "..").
If the path is valid, it will stream the file content directly to the HTTP response with proper headers values
for file download. Errors during streaming are logged, but cannot be reported to the client once the response
has started. 

Absolute path traversal is prevented to ensure that users cannot specify arbitrary locations on the server's
filesystem, which could lead to unauthorized access to sensitive files outside the application's intended
directory. By restricting to relative paths and validating them, the function mitigates security risks
associated with directory traversal and file disclosure vulnerabilities.

File: api/internal/features/file-manager/controller/filename_utils.go
Comment (yashksaini-coder) [reply api/internal/features/file-manager/controller/filename_utils.go:1]: Split into small functions

File: api/internal/features/file-manager/controller/upload_file.go (hunk 3 above)
Comment (yashksaini-coder) [reply api/internal/features/file-manager/controller/upload_file.go:31]: update to reduce the nested if conditions.

**Comment by yashksaini-coder** [comment]
Closed this cause I break the code and tests failed to build. Will re-raise in near future.
